File: android/app/src/main/java/sol/vortexx/android/identity/impl/SeedIdentityRepository.kt
package sol.vortexx.android.identity.impl

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import sol.vortexx.android.auth.api.SecureStore
import sol.vortexx.android.crypto.api.KeyPair
import sol.vortexx.android.crypto.api.Kdf
import sol.vortexx.android.crypto.util.Hex
import sol.vortexx.android.identity.api.Identity
import sol.vortexx.android.identity.api.IdentityRepository
import sol.vortexx.android.identity.api.Mnemonic
import sol.vortexx.android.identity.api.SeedProvider
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Derives and persists the device identity.
 *
 * On first call: generate mnemonic → derive 64-byte seed → HKDF split
 * into an X25519 seed and an Ed25519 seed → generate the two keypairs.
 * Mnemonic + private keys land in [SecureStore]; public keys are
 * exported to non-secure DataStore (Wave 8+) so lookup doesn't touch
 * the Keystore every frame.
 */
@Singleton
class SeedIdentityRepository @Inject constructor(
    private val seeds: SeedProvider,
    private val kdf: Kdf,
    private val store: SecureStore,
) : IdentityRepository {

    private val _identity = MutableStateFlow<Identity?>(loadFromStore())
    override val identity = _identity.asStateFlow()

    override suspend fun createOrLoad(): Identity {
        _identity.value?.let { return it }

        // Consult SecureStore again — an out-of-band restore (e.g. key-backup
        // flow) can have populated it after wipe(). Without this step we'd
        // generate a brand new mnemonic and overwrite the restored one.
        loadFromStore()?.let {
            _identity.value = it
            return it
        }

        val mnemonic = seeds.generate()
        val seed = seeds.toSeed(mnemonic)
        val id = deriveIdentity(mnemonic, seed)
        persist(id)
        _identity.value = id
        return id
    }

    override suspend fun wipe() {
        store.putString(KEY_MNEMONIC, null)
        store.putString(KEY_X25519_PRIV, null)
        store.putString(KEY_X25519_PUB,  null)
        store.putString(KEY_ED25519_PRIV, null)
        store.putString(KEY_ED25519_PUB,  null)
        _identity.value = null
    }

    // ── internals ──────────────────────────────────────────────────────

    private fun deriveIdentity(mnemonic: Mnemonic, seed: ByteArray): Identity {
        // Domain-separated HKDF: one derivation per key role so rotating
        // one of them doesn't leak the other.
        val xSeed = kdf.derive(ikm = seed, info = "vortex/x25519".toByteArray(), length = 32)
        val eSeed = kdf.derive(ikm = seed, info = "vortex/ed25519".toByteArray(), length = 32)

        val xPriv = X25519PrivateKeyParameters(xSeed, 0)
        val ePriv = Ed25519PrivateKeyParameters(eSeed, 0)
        return Identity(
            mnemonic = mnemonic,
            x25519  = KeyPair(privateKey = xPriv.encoded.copyOf(), publicKey = xPriv.generatePublicKey().encoded.copyOf()),
            ed25519 = KeyPair(privateKey = ePriv.encoded.copyOf(), publicKey = ePriv.generatePublicKey().encoded.copyOf()),
        )
    }

    private fun persist(id: Identity) {
        store.putString(KEY_MNEMONIC,      id.mnemonic.toString())
        store.putString(KEY_X25519_PRIV,   Hex.encode(id.x25519.privateKey))
        store.putString(KEY_X25519_PUB,    Hex.encode(id.x25519.publicKey))
        store.putString(KEY_ED25519_PRIV,  Hex.encode(id.ed25519.privateKey))
        store.putString(KEY_ED25519_PUB,   Hex.encode(id.ed25519.publicKey))
    }

    private fun loadFromStore(): Identity? {
        val words = store.getString(KEY_MNEMONIC) ?: return null
        return runCatching {
            Identity(
                mnemonic = Mnemonic(words.split(' ')),
                x25519  = KeyPair(
                    privateKey = Hex.decode(store.getString(KEY_X25519_PRIV)!!),
                    publicKey  = Hex.decode(store.getString(KEY_X25519_PUB )!!),
                ),
                ed25519 = KeyPair(
                    privateKey = Hex.decode(store.getString(KEY_ED25519_PRIV)!!),
                    publicKey  = Hex.decode(store.getString(KEY_ED25519_PUB )!!),
                ),
            )
        }.getOrNull()
    }

    private companion object {
        const val KEY_MNEMONIC      = "id_mnemonic"
        const val KEY_X25519_PRIV   = "id_x25519_priv"
        const val KEY_X25519_PUB    = "id_x25519_pub"
        const val KEY_ED25519_PRIV  = "id_ed25519_priv"
        const val KEY_ED25519_PUB   = "id_ed25519_pub"
    }
}

File: ios/Modules/Sources/Identity/impl/SeedIdentityRepository.swift
import Foundation
import CryptoKit
import VortexCrypto
import Auth

/// Derives and persists the device identity.
///
/// On first call: generate mnemonic → derive 64-byte BIP39 seed → HKDF
/// split into X25519 + Ed25519 32-byte seeds → generate key pairs from
/// them → persist to Keychain via [SecureStore].
public final class SeedIdentityRepository: IdentityRepository, @unchecked Sendable {
    private let seeds: SeedProvider
    private let kdf: Kdf
    private let store: SecureStore
    private let lock = NSLock()
    private var cached: Identity?

    public init(seeds: SeedProvider, kdf: Kdf, store: SecureStore) {
        self.seeds = seeds
        self.kdf = kdf
        self.store = store
        self.cached = SeedIdentityRepository.loadFromStore(store)
    }

    public var current: Identity? {
        lock.lock(); defer { lock.unlock() }
        return cached
    }

    public func createOrLoad() async throws -> Identity {
        if let c = current { return c }
        // Consult store again — an external restore (Wave 20 key-backup
        // flow) may have populated it after wipe().
        if let fromStore = SeedIdentityRepository.loadFromStore(store) {
            lock.lock(); cached = fromStore; lock.unlock()
            return fromStore
        }
        let mnemonic = try seeds.generate()
        let seed = try seeds.toSeed(mnemonic: mnemonic)
        let id = try derive(mnemonic: mnemonic, seed: seed)
        persist(id)
        lock.lock(); cached = id; lock.unlock()
        return id
    }

    public func wipe() {
        store.setString(Keys.mnemonic,      nil)
        store.setString(Keys.x25519Priv,    nil)
        store.setString(Keys.x25519Pub,     nil)
        store.setString(Keys.ed25519Priv,   nil)
        store.setString(Keys.ed25519Pub,    nil)
        lock.lock(); cached = nil; lock.unlock()
    }

    // ── internals ──────────────────────────────────────────────────────

    private func derive(mnemonic: Mnemonic, seed: Data) throws -> Identity {
        let xSeed = try kdf.derive(ikm: seed, salt: Data(), info: Data("vortex/x25519".utf8), length: 32)
        let eSeed = try kdf.derive(ikm: seed, salt: Data(), info: Data("vortex/ed25519".utf8), length: 32)

        let xPriv = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: xSeed)
        let ePriv = try Curve25519.Signing.PrivateKey(rawRepresentation: eSeed)
        return Identity(
            mnemonic: mnemonic,
            x25519: KeyPair(privateKey: xPriv.rawRepresentation,
                            publicKey:  xPriv.publicKey.rawRepresentation),
            ed25519: KeyPair(privateKey: ePriv.rawRepresentation,
                             publicKey:  ePriv.publicKey.rawRepresentation),
        )
    }

    private func persist(_ id: Identity) {
        store.setString(Keys.mnemonic,      id.mnemonic.phrase)
        store.setString(Keys.x25519Priv,    Hex.encode(id.x25519.privateKey))
        store.setString(Keys.x25519Pub,     Hex.encode(id.x25519.publicKey))
        store.setString(Keys.ed25519Priv,   Hex.encode(id.ed25519.privateKey))
        store.setString(Keys.ed25519Pub,    Hex.encode(id.ed25519.publicKey))
    }

    private static func loadFromStore(_ s: SecureStore) -> Identity? {
        guard let phrase = s.getString(Keys.mnemonic) else { return nil }
        let parts = phrase.split(separator: " ").map(String.init)
        guard parts.count == 24,
              let xPriv = try? Hex.decode(s.getString(Keys.x25519Priv) ?? ""),
              let xPub  = try? Hex.decode(s.getString(Keys.x25519Pub)  ?? ""),
              let ePriv = try? Hex.decode(s.getString(Keys.ed25519Priv) ?? ""),
              let ePub  = try? Hex.decode(s.getString(Keys.ed25519Pub)  ?? ""),
              let m = try? Mnemonic(words: parts)
        else { return nil }
        return Identity(
            mnemonic: m,
            x25519:  KeyPair(privateKey: xPriv, publicKey: xPub),
            ed25519: KeyPair(privateKey: ePriv, publicKey: ePub),
        )
    }

    private enum Keys {
        static let mnemonic    = "id_mnemonic"
        static let x25519Priv  = "id_x25519_priv"
        static let x25519Pub   = "id_x25519_pub"
        static let ed25519Priv = "id_ed25519_priv"
        static let ed25519Pub  = "id_ed25519_pub"
    }
}
